### PR TITLE
Efficient CNN

### PR DESCRIPTION
**Pull request by nsorros**
Description
---

This PR reduces the memory requirement needed for CNN by introducing the ability to pass a tensorflow dataset as an input. In order to unify how data are handled within CNN, when data passed is a list of numpy array, they are converted to a tensorflow dataset as well.

What makes tensorflow dataset light in memory is their ability to initialise from a generator. Actually passing a generator is another way CNN can become memory light but then functionality around batching, shuffling etc would need to be implemented when they come for free with a tensorflow dataset.

Checklist
---

- [x] Added link to Github issue or Trello card
- [x] Added tests

### CODE REVIEW

**Comment by nsorros** [comment]
[Refactor CNN and train to be memory efficient](https://trello.com/c/pDkyOynd/1175-refactor-cnn-and-train-to-be-memory-efficient)

File: tests/test_cnn.py [hunk 1]
```diff
@@ -151,12 +159,14 @@ def test_early_stopping():
...
         ('vec', KerasVectorizer()),
         ('clf', CNNClassifier(
+                    batch_size=2,
                     early_stopping=True,
                     nb_epochs=10000
         ))
     ])
     # if early_stopping is not working it will take
     # a lot of time to finish running this test
+    # it will also consume the 4MB of logs in travis
```
Comment (aCampello) [inline tests/test_cnn.py:169]: Can we define that env variable to silence the logs at least in travis (https://github.com/wellcometrust/WellcomeML/issues/209) ?

File: wellcomeml/ml/cnn.py [hunk 2]
```diff
@@ -98,9 +112,53 @@ def _get_distributed_strategy(self):
             strategy = tf.distribute.get_strategy()
         return strategy
 
+    def _init_from_data(self, X, Y):
+        logger.info(
+            "Initializing sequence_length, vocab_size \
+            and nb_outputs from data. This might take a while."
+        )
+        if type(X) is np.ndarray:
```
Comment (aCampello) [inline wellcomeml/ml/cnn.py:120]: `isinstance(X, np.ndarray)` will do too

File: wellcomeml/ml/cnn.py [hunk 3]
```diff
@@ -98,9 +112,53 @@ def _get_distributed_strategy(self):
...
+            X = X.batch(self.batch_size)
+
+            # init from iterating over dataset
+            X_max = 0
+            Y_max = 0
+            X_shape = None
+            Y_shape = None
+            steps_per_epoch = 0
+            for X_batch, Y_batch in X:
+                batch_X_max = X_batch.numpy().max()
```
Comment (aCampello) [inline wellcomeml/ml/cnn.py:136]: I don't understand this `.numpy().max()` operation..

Comment (aCampello) [reply wellcomeml/ml/cnn.py:136]: Does this give you the length? I thought that's what you needed to calculate

**Review by aCampello — APPROVED** [review]
Took me a while to understand the logic between `.take` and `.skip`, but all in all it's great work! 👍 If you feel like, could put one of the tests into `/example`.

File: tests/test_cnn.py (hunk 1 above)
Comment (nsorros) [reply tests/test_cnn.py:169]: Good point will do that

File: wellcomeml/ml/cnn.py (hunk 2 above)
Comment (nsorros) [reply wellcomeml/ml/cnn.py:120]: possibly better and more consistent, will dot the change

File: wellcomeml/ml/cnn.py (hunk 3 above)
Comment (nsorros) [reply wellcomeml/ml/cnn.py:136]: it gives me the maximum integer in X batch. I want to find the vocabulary size.

the length of the columns is also used to define sequence length but the variable is called X_shape.

happy to rename if anything is unclear

File: wellcomeml/ml/cnn.py [hunk 4]
```diff
@@ -180,37 +240,48 @@ def residual_attention(x1):
...
+            steps_per_epoch = self._init_from_data(X, Y)
+
+        if type(X) is np.ndarray:
+            data = self._prepare_data(X, Y, shuffle=True)
+        else:  # tensorflow dataset
+            data = X.batch(self.batch_size)
+
+        train_steps_per_epoch = int((1-self.validation_split) * steps_per_epoch)
+        if train_steps_per_epoch == 0:
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (nsorros) [reply wellcomeml/ml/cnn.py:272]: it still works but there are no validation metrics. actually early stopping does not work so maybe I should mention

File: wellcomeml/ml/cnn.py (hunk 2 above)
Comment (nsorros) [reply wellcomeml/ml/cnn.py]: b9bf570

File: wellcomeml/ml/cnn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nsorros) [reply wellcomeml/ml/cnn.py:272]: 5ae533c

**Pull request merged**